dao/src/main/java/com/epam/esm/dao/parameters/impl/SearchQueryProviderImpl.java

```diff
@@ -1,69 +1,98 @@
+import com.epam.esm.domain.util.SearchParameter;
 import org.springframework.stereotype.Component;
 
+import javax.persistence.EntityManager;
+import javax.persistence.criteria.*;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static com.epam.esm.dao.util.ParameterValue.PARAMETER_NAME;
+import static com.epam.esm.dao.util.ParameterValue.PARAMETER_TAG;
 
 @Component
 public class SearchQueryProviderImpl implements SearchQueryProvider {
 
-    private static final String PART_WHERE = "WHERE ";
-    private static final String PART_HAS_TAG = "t.name = ? AND ";
-    private static final String PART_NAME_LIKE = "c.name LIKE ";
-    private static final String PARAMETER_PREPARED_LIKE = "concat('%', ?, '%') ";
-    private static final String PARAMETER_LIKE_ANY = "'%%' ";
-    private static final String PART_DESCRIPTION_LIKE = "AND c.description LIKE ";
-    private static final String ORDER_BY_CERTIFICATE = "ORDER BY c.";
-
-    private static final String NAME = "name";
-    private static final String SORT_TYPE_DEFAULT = "create_date ";
-    private static final String SORT_TYPE_NAME = "name ";
-    private static final String ORDER_TYPE_DEFAULT = "ASC";
+    private static final String PARAMETER_TAG_NAMES = "tagNames";
+    private static final String SELECT_TAG_BY_NAME = "from Tag tag where tag.name in (:" + PARAMETER_TAG_NAMES + ")";
+    private static final String PARAMETER_ANY = "%";
+    private static final String SORT_TYPE_DEFAULT = "createDate";
     private static final String ORDER_TYPE_DESC = "DESC";
+    private static final String ATTRIBUTE_DESCRIPTION = "description";
+
+    private final EntityManager entityManager;
 
-    @Override
-    public String getCertificateSearchQuery(String tagName, String certificateName, String description, String sortBy, String order) {
-        return tagQuery(tagName)
-                .concat(nameQuery(certificateName))
-                .concat(descriptionQuery(description))
-                .concat(sortOrDefault(sortBy))
-                .concat(orderOrDefault(order));
+    public SearchQueryProviderImpl(EntityManager entityManager) {
+        this.entityManager = entityManager;
     }
 
-    private String tagQuery(String tagName) {
-        if (Objects.nonNull(tagName)) {
-            return PART_WHERE.concat(PART_HAS_TAG);
-        }
-        return PART_WHERE;
+    private CriteriaBuilder criteriaBuilder;
+    private CriteriaQuery<Certificate> criteriaQuery;
+    private Root<Certificate> certificateRoot;
+
+
+    public CriteriaQuery<Certificate> buildQuery(SearchParameter parameter) {
+        criteriaBuilder = entityManager.getCriteriaBuilder();
+        criteriaQuery = criteriaBuilder.createQuery(Certificate.class);
+        certificateRoot = criteriaQuery.from(Certificate.class);
+
+        // set search predicates
+        List<Predicate> predicateList = buildPredicateList(parameter);

```

Comment:
If you have intention to place a comment or empty place to improve readability (or even both of them), it means that it's time to split the logic to several methods and name the appropriately. The code should be descriptive itself.

Reply:
thank you for comment. refactor performed